ampd/src/lib.rs

```diff
@@ -275,7 +281,34 @@ where
         };
 
         while (set.join_next().await).is_some() {}
+        let state_update_result = state_rx.try_recv().change_context(Error::ReturnState);
 
-        res
+        match (execution_result, state_update_result) {
+            (Err(mut err1), Err(err2)) => {
+                err1.extend_one(err2);
+                Err(err1)
+            }
+            (Err(err1), Ok(_)) => Err(err1),
+            (Ok(_), Err(err2)) => Err(err2),
```

Comment:
We still need to return the latest state to persist though even in the case of `execution_result` being an error.